<a href="https://colab.research.google.com/github/Matheesha-w/SE4050-Lab6/blob/main/NetworkX_tutorial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **SE4050 - Lab 8**

 We highly recommend you to run this notebook, so you can get familiar with the basic concepts of graphs and networks.

In this Colab, we will introduce a graph related package, [NetworkX](https://networkx.org/documentation/stable/).

NetworkX is one of the most frequently used Python packages to create, manipulate, and mine graphs.

Main parts of this tutorial are adapted from https://colab.research.google.com/github/jdwittenauer/ipython-notebooks/blob/master/notebooks/libraries/NetworkX.ipynb#scrollTo=zA1OO6huHeV6

Please make a copy before you proceed.

In [ ]:
# Upgrade packages
!pip install --upgrade scipy networkx

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 24.2 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [ ]:
# Import the NetworkX package
import networkx as nx
import numpy as np
import matplotlib.pyplot as plt

## Graph
NetworkX provides several classes to store different types of graphs, such as directed and undirected graph. It also provides classes to create multigraphs (both directed and undirected).

For more information, please refer to [NetworkX graph types](https://networkx.org/documentation/stable/reference/classes/index.html).

In [ ]:
# Create an undirected graph G
G = nx.Graph()
print(G.is_directed())

# Create a directed graph H
H = nx.DiGraph()
print(H.is_directed())

# Add graph level attribute
G.graph["Name"] = "Bar"
print(G.graph)

False
True
{'Name': 'Bar'}


## Node

Nodes (with attributes) can be easily added to NetworkX graphs.

In [ ]:
# Add one node with node level attributes
G.add_node(0, feature=5, label=0)

# Get attributes of the node 0
node_0_attr = G.nodes[0]
print("Node 0 has the attributes {}".format(node_0_attr))

Node 0 has the attributes {'feature': 5, 'label': 0}


In [ ]:
G.nodes(data=True)

NodeDataView({0: {'feature': 5, 'label': 0}})

In [ ]:
# Add multiple nodes with attributes
G.add_nodes_from([
  (1, {"feature": 1, "label": 1}),
  (2, {"feature": 2, "label": 2})
]) #(node, attrdict)

# Loop through all the nodes
# Set data=True will return node attributes
for node in G.nodes(data=True):
  print(node)

# Get number of nodes
num_nodes = G.number_of_nodes()
print("G has {} nodes".format(num_nodes))

(0, {'feature': 5, 'label': 0})
(1, {'feature': 1, 'label': 1})
(2, {'feature': 2, 'label': 2})
G has 3 nodes


## Edge

Similar to nodes, edges (with attributes) can also be easily added to NetworkX graphs.

In [ ]:
# Add one edge with edge weight 0.5
G.add_edge(0, 1, weight=0.5)

# Get attributes of the edge (0, 1)
edge_0_1_attr = G.edges[(0, 1)]
print("Edge (0, 1) has the attributes {}".format(edge_0_1_attr))

Edge (0, 1) has the attributes {'weight': 0.5}


In [ ]:
# Add multiple edges with edge weights
G.add_edges_from([
  (1, 2, {"weight": 0.3}),
  (2, 0, {"weight": 0.1})
])

# Loop through all the edges
# Here there is no data=True, so only the edge will be returned
for edge in G.edges():
  print(edge)

# Get number of edges
num_edges = G.number_of_edges()
print("G has {} edges".format(num_edges))

(0, 1)
(0, 2)
(1, 2)
G has 3 edges


## Visualization

In [ ]:
# Draw the graph
nx.draw(G, with_labels = True)

## Node Degree and Neighbor

In [ ]:
node_id = 1

# Degree of node 1
print("Node {} has degree {}".format(node_id, G.degree[node_id]))

# Get neighbor of node 1
for neighbor in G.neighbors(node_id):
  print("Node {} has neighbor {}".format(node_id, neighbor))

Node 1 has degree 2
Node 1 has neighbor 0
Node 1 has neighbor 2


## Random Graph creation and analysis

In [ ]:
# creation of a random network with N nodes and E edges - this a erdos-renyi graph
def create(n,m):
    G = nx.gnm_random_graph(n, m, seed = random_state)
    return G

In [ ]:
def degree_analysis(G):
    degree_sequence = sorted((d for n, d in G.degree()), reverse=True)

    fig = plt.figure("Degree of a random graph", figsize=(8, 8))
    # Create a gridspec for adding subplots of different sizes
    axgrid = fig.add_gridspec(2, 4)


    ax1 = fig.add_subplot(axgrid[:, :2])
    ax1.plot(degree_sequence, "b-", marker="o")
    ax1.set_title("Degree Rank Plot")
    ax1.set_ylabel("Degree")
    ax1.set_xlabel("Rank")

    ax2 = fig.add_subplot(axgrid[:, 2:])
    ax2.bar(*np.unique(degree_sequence, return_counts=True))
    ax2.set_title("Degree histogram")
    ax2.set_xlabel("Degree")
    ax2.set_ylabel("# of Nodes")

    plt.show()

In [ ]:
random_state = np.random.RandomState(42)
N=20 #number of nodes
E=2*N #number of edges
G=create(N,E)

In [ ]:
print(G.edges(nbunch=4))
print(G.adj)
print(G.edges)
print(type(G))
print(len(G.edges))

[(4, 18), (4, 9)]
{0: {11: {}, 12: {}, 2: {}, 1: {}}, 1: {11: {}, 5: {}, 19: {}, 9: {}, 0: {}}, 2: {7: {}, 19: {}, 14: {}, 0: {}}, 3: {10: {}, 13: {}, 17: {}, 7: {}, 9: {}}, 4: {18: {}, 9: {}}, 5: {1: {}}, 6: {19: {}, 7: {}, 8: {}, 17: {}, 14: {}}, 7: {6: {}, 2: {}, 11: {}, 3: {}, 13: {}, 15: {}, 10: {}}, 8: {6: {}, 17: {}, 14: {}}, 9: {15: {}, 3: {}, 1: {}, 4: {}}, 10: {14: {}, 18: {}, 3: {}, 7: {}}, 11: {1: {}, 0: {}, 16: {}, 18: {}, 7: {}, 17: {}}, 12: {17: {}, 14: {}, 0: {}}, 13: {3: {}, 16: {}, 7: {}}, 14: {10: {}, 6: {}, 2: {}, 15: {}, 12: {}, 8: {}}, 15: {9: {}, 14: {}, 7: {}}, 16: {11: {}, 13: {}, 18: {}}, 17: {6: {}, 8: {}, 3: {}, 11: {}, 12: {}}, 18: {10: {}, 11: {}, 4: {}, 16: {}}, 19: {6: {}, 2: {}, 1: {}}}
[(0, 11), (0, 12), (0, 2), (0, 1), (1, 11), (1, 5), (1, 19), (1, 9), (2, 7), (2, 19), (2, 14), (3, 10), (3, 13), (3, 17), (3, 7), (3, 9), (4, 18), (4, 9), (6, 19), (6, 7), (6, 8), (6, 17), (6, 14), (7, 11), (7, 13), (7, 15), (7, 10), (8, 17), (8, 14), (9, 15), (10, 14), 

In [ ]:
pos_cons = nx.spring_layout(G,seed = random_state)
nx.draw(G, pos=pos_cons,labels ={node:node for node in G.nodes()}) #spring_layout , spectral_layout
plt.draw()


In [ ]:
# Get the adjacency matrix
A = nx.adjacency_matrix(G).toarray()
print("Adjacency Matrix:")
print(A)



# Get the degree matrix
degree_sequence_full = [d for n, d in G.degree()]
D = np.diag(degree_sequence_full)
print("\nDegree Matrix:")
print(D)



# Get the Laplacian matrix
L = D - A
print("\nLaplacian Matrix:")
print(L)


Adjacency Matrix:
[[0 1 1 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0]
 [1 0 0 0 0 1 0 0 0 1 0 1 0 0 0 0 0 0 0 1]
 [1 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 1]
 [0 0 0 0 0 0 0 1 0 1 1 0 0 1 0 0 0 1 0 0]
 [0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0]
 [0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 1 1 0 0 0 0 0 1 0 0 1 0 1]
 [0 0 1 1 0 0 1 0 0 0 1 1 0 1 0 1 0 0 0 0]
 [0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 1 0 0]
 [0 1 0 1 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0]
 [0 0 0 1 0 0 0 1 0 0 0 0 0 0 1 0 0 0 1 0]
 [1 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 1 1 0]
 [1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0]
 [0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0]
 [0 0 1 0 0 0 1 0 1 0 1 0 1 0 0 1 0 0 0 0]
 [0 0 0 0 0 0 0 1 0 1 0 0 0 0 1 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 0 1 0]
 [0 0 0 1 0 0 1 0 1 0 0 1 1 0 0 0 0 0 0 0]
 [0 0 0 0 1 0 0 0 0 0 1 1 0 0 0 0 1 0 0 0]
 [0 1 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0]]

Degree Matrix:
[[4 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 5 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
 [0 0 4 0 0 0 0 0 0

In [ ]:
degree_analysis(G)

C:\Users\Matheesha\AppData\Local\Temp\ipykernel_8992\2905163198.py:21: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


In [ ]:
# calculate the graph density for the above created random graph
V = G.number_of_nodes()
E_count = G.number_of_edges()
density = (2 * E_count) / (V * (V - 1))
print(f"Number of nodes |V| = {V}")
print(f"Number of edges |E| = {E_count}")
print(f"Graph density D = {density:.5f}")


Number of nodes |V| = 20
Number of edges |E| = 40
Graph density D = 0.21053


In [ ]:
# increase N from 20 to 200 and see how density and degree distribution change
N_values = [20, 40, 60, 80, 100, 120, 150, 180, 200]
densities = []
mean_degrees = []

fig, axes = plt.subplots(3, 3, figsize=(16, 14))
axes = axes.flatten()

for i, n in enumerate(N_values):
    random_state = np.random.RandomState(42)
    m = 2 * n  # keep E = 2N like the original graph
    Gn = create(n, m)

    d = (2 * Gn.number_of_edges()) / (n * (n - 1))
    densities.append(d)

    deg_seq = [deg for node, deg in Gn.degree()]
    mean_degrees.append(np.mean(deg_seq))

    axes[i].bar(*np.unique(deg_seq, return_counts=True))
    axes[i].set_title(f"N={n}, density={d:.4f}")
    axes[i].set_xlabel("Degree")
    axes[i].set_ylabel("# of Nodes")

plt.tight_layout()
plt.show()

plt.figure(figsize=(6, 4))
plt.plot(N_values, densities, marker='o')
plt.xlabel("N (number of nodes)")
plt.ylabel("Graph density D")
plt.title("Graph density vs N")
plt.show()

for n, d, md_ in zip(N_values, densities, mean_degrees):
    print(f"N={n:4d}  density={d:.5f}  mean_degree={md_:.2f}")


N=  20  density=0.21053  mean_degree=4.00
N=  40  density=0.10256  mean_degree=4.00
N=  60  density=0.06780  mean_degree=4.00
N=  80  density=0.05063  mean_degree=4.00
N= 100  density=0.04040  mean_degree=4.00
N= 120  density=0.03361  mean_degree=4.00
N= 150  density=0.02685  mean_degree=4.00
N= 180  density=0.02235  mean_degree=4.00
N= 200  density=0.02010  mean_degree=4.00


C:\Users\Matheesha\AppData\Local\Temp\ipykernel_8992\304903218.py:26: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()
C:\Users\Matheesha\AppData\Local\Temp\ipykernel_8992\304903218.py:33: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


As N increases from 20 to 200, the density drops a lot (from about 0.21 down to 0.02), even though the average degree stays about the same (~4) since we kept E = 2N. This makes sense because density depends on N^2 in the denominator, so it shrinks fast as the graph gets bigger, even if each node still has roughly the same number of neighbours.

The degree histogram also looks more like a smooth bell curve at N=200 compared to N=20, where it's more random/uneven just because there are fewer nodes to average over.
